rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+
+#### Private key policy discovery
+
+Teleport clients should be able to automatically determine if a user requires a hardware private key for login to avoid additional UX concerns. For this, we will introduce the `GetPrivateKeyPolicy` rpc.
+
+```proto
+service AuthService {
+  rpc GetPrivateKeyPolicy(GetPrivateKeyPolicyRequest) returns (GetPrivateKeyPolicyResponse);
+}
+
+// The user will be pulled from the request certificate, so request message can be empty for now.
+message GetPrivateKeyPolicyRequest {}
+
+message GetPrivateKeyPolicyResponse {
+    PrivateKeyPolicy policy = 1;
+}
+
+enum PrivateKeyPolicy {
+    PRIVATE_KEY_POLICY_NONE = 0;
+    PRIVATE_KEY_POLICY_HARDWARE_KEY = 1;
+    PRIVATE_KEY_POLICY_HARDWARE_KEY_TOUCH = 2;
+}
+```
+
+In order for a Teleport Client to call `GetPrivateKeyPolicy`, it first needs a set of valid certificates. Unlike `IsMFARequired`, which issues an MFA certificate Reissue request after logging in with non-MFA certificates, we need to call `GetPrivateKeyPolicy` on the initial login.
+
+One option is to make add an endpoint in the Proxy Web API which can be called before login. In order to authenticate the user, we would need them to login with their login credentials. For MFA and Password login, we can potentially reuse these password credentials to intiate the actual login directly after retrieving the user's private key policy requirement. However, for sso login and some edge cases is MFA login (OTP token timing, between requests, etc.), the user would be required to perform the login twice.
+
+Alternatively, we could provide a user with set of certificates using a plain RSA private key in memory first. The resulting certificates will fail the hardware key attestation check in the Auth server, so they will be invalid for all API requests, except for `GetPrivateKeyPolicy` and `GenerateUserCerts`. This way, after the initial login, `GenerateUserCerts` can be used to issue new certificates with a hardware private key.
+
+While this strategy does prevent the user from needing to login twice, it presents a possible workaround. If a Teleport Client stores the initial RSA private key and certificates, then they can be used by anyone to reissue hardware key certifificates. To mitigate this risk, the initial certificates will have a TTL of 1 minute and will only be authorized for a single reissue request. This flow is similar to per-session MFA, which also issues one-time-use certificates with a 1 minute TTL.
+
+To avoid situtations where the user has to login twice, we will go with the second option.
+
+#### Hardware private key login
+
+As mentioned above, hardware key login will start with the normal login flow, followed by a call to `GetPrivateKeyPolicy`. If the result is `PRIVATE_KEY_POLICY_HARDWARE_KEY` or `PRIVATE_KEY_POLICY_HARDWARE_KEY_TOUCH`, then the Teleport client will:
+ 1. Find a hardware key on the user's device
```

Comment:
Could we proactively look for keys? Won't that save us from some of the issues that arise from the delayed check?